kasa/cli.py

```diff
@@ -630,18 +631,113 @@ async def schedule(dev):
-    for rule in sched.rules:
-        print(rule)
-    else:
+    if sched.rules == []:
         echo(f"No rules of type {type}")
+    for rule in sched.rules:
+        if json:
+            print(rule.json())
+        else:
+            print(rule)
 
     return sched.rules
 
 
+@schedule.command(name="enable")
+@pass_dev
+@click.argument("enable", type=click.BOOL)
+async def _schedule_enable(dev, enable):
+    """Enable or disable schedule."""
+    schedule = dev.modules["schedule"]
+    return await schedule.set_enabled(1 if state else 0)
+
+
+@schedule.command(name="add")
+@pass_dev
+@click.option("--name", type=str, required=True)
+@click.option("--enable", type=click.BOOL, default=True, show_default=True)
+@click.option("--repeat", type=click.BOOL, default=True, show_default=True)
+@click.option("--wday", type=str, required=True)
+@click.option("--sact", type=click.IntRange(-1, 2), default=None, required=True)
+@click.option("--stime_opt", type=click.IntRange(-1, 2), default=None, required=True)
+@click.option("--smin", type=click.IntRange(0, 1440), default=None, required=True)
+@click.option("--eact", type=click.IntRange(-1, 2), default=-1)
+@click.option("--etime_opt", type=click.IntRange(-1, 2), default=-1)
+@click.option("--emin", type=click.IntRange(0, 1440), default=None)
+async def add_rule(
+    dev, name, enable, repeat, wday, sact, stime_opt, smin, eact, etime_opt, emin
+):
+    """Add rule to device."""
+    schedule = dev.modules["schedule"]
+    rule_to_add = Rule(
+        name=name,
+        enable=enable,
+        repeat=repeat,
+        wday=list(map(int, wday.split(","))),
+        sact=sact,
+        stime_opt=stime_opt,
+        smin=smin,
+        eact=eact,
+        etime_opt=etime_opt,
+        emin=emin,
+    )
+    if rule_to_add:
+        echo("Adding rule")
+        return await schedule.add_rule(rule_to_add)
+    else:
+        echo("Invalid rule")
+
+
+@schedule.command(name="edit")
+@pass_dev
+@click.option("--id", type=str, required=True)
+@click.option("--name", type=str)
+@click.option("--enable", type=click.BOOL)
+@click.option("--repeat", type=click.BOOL)
+@click.option("--wday", type=str)
+@click.option("--sact", type=click.IntRange(-1, 2))
+@click.option("--stime_opt", type=click.IntRange(-1, 2))
+@click.option("--smin", type=click.IntRange(0, 1440))
+@click.option("--eact", type=click.IntRange(-1, 2))
+@click.option("--etime_opt", type=click.IntRange(-1, 2))
+@click.option("--emin", type=click.IntRange(0, 1440))
+async def edit_rule(
+    dev, id, name, enable, repeat, wday, sact, stime_opt, smin, eact, etime_opt, emin
+):
+    """Edit rule from device."""
+    schedule = dev.modules["schedule"]
+    rule_to_edit = next(filter(lambda rule: (rule.id == id), schedule.rules), None)
+    if rule_to_edit:
+        echo(f"Editing rule id {id}")
+        if name is not None:
+            rule_to_edit.name = name
+        if enable is not None:
+            rule_to_edit.enable = 1 if enable else 0
```

Comment:
pydantic should handle the conversion just fine as long as the types are correct, so there should be no need for manual conversions here.

Reply:
I want that only the user specified options should override the existing rule fields, however if the user does not specify a flag we get a `None` value that i don't want to pass to pydantic to overwrite the field. If there is a better way to do this with click & pydantic please let me know. (I am not too familiar with either of them)